Issue #13603 - Bad Accept-Language parsing fix for Locale. [12.0.x]

joakime opened the pull request:
If a value-less entry in an `Accept-Language` request header is received, this can result in a failure when looking up the Locale of the request.

Example bad Request Headers.

```
Accept-Language: ;q=0.5
Accept-Language: q=0.6
```

This actually results in the QuotedQualityCSV in having an entry that is blank.
The fix in this PR fixes the parsing of Quoted CSV headers to ignore those headers where the quality exists without a value, not adding them to the sorted list of values.

Fixes: #13603

joakime committed:
Issue #13603 - Bad Accept-Language parsing fix for Locale.

- changes to HttpServletRequest.getLocale() to filter bad locale entries
  (blank, not-valid, etc) from results of HttpField
  quoted-quality-csv parsing.

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/QuotedCSV.java:
```diff
@@ -118,7 +118,7 @@ public QuotedCSV(boolean keepQuotes, String... values)
 
     @Override
-    protected void parsedValueAndParams(StringBuilder buffer)
+    protected void parsedValueAndParams(StringBuilder buffer, int valueLength)
```
why is valueLength ignored?

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/QuotedCSVParser.java:
```diff
@@ -175,7 +175,7 @@ else if (state == State.PARAM_VALUE && paramValue < 0)
                         }
 
-                        parsedValueAndParams(buffer);
+                        parsedValueAndParams(buffer, valueLength);
```
The value length is just the buffer length and it is ignored anyway?

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/QuotedQualityCSV.java:
```diff
@@ -134,6 +139,10 @@ protected void parsedValueAndParams(StringBuilder buffer)
     protected void parsedValue(StringBuilder buffer)
     {
+        // ignore blank/invalid values
+        if (buffer == null || StringUtil.isBlank(buffer.toString()))
```
The whole point of a method taking StringBuilder is to avoid creating a string in toString

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/QuotedQualityCSV.java:
```diff
@@ -122,9 +123,13 @@ public QuotedQualityCSV(ToIntFunction<String> secondaryOrdering)
     {
-        super.parsedValueAndParams(buffer);
+        // ignore blank/invalid values
+        if (valueLength <= 0 || buffer == null || StringUtil.isBlank(buffer.toString()))
```
The whole point of a method taking StringBuilder is to avoid creating a string in toString
Plus this ignores valueLength too

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/QuotedQualityCSV.java:
```diff
@@ -147,6 +156,10 @@ protected void parsedValue(StringBuilder buffer)
     protected void parsedParam(StringBuilder buffer, int valueLength, int paramName, int paramValue)
     {
+        // ignore blank/invalid values
+        if (valueLength <= 0 || buffer == null || StringUtil.isBlank(buffer.toString()))
```
ditto

gregw reviewed (changes requested):
The value length arg is not needed and is mostly ignored.
do not call toString in methods passed the StringBuilder as the intention is to avoid creating a string in the first place

joakime commented:
@gregw I reworked some of the points you highlighted with regards to buffer.toString() and valueLength.

The `buffer.toString()` is required for the `new QualityValue()` call later, so I moved the `buffer.toString()` to a more appropriate point.

Regarding the `valueLength` parameter, it has value to identify bad values (not bad parameters).
This is used to prevent the `new QualityValue()` call from creating a bogus entry with an empty value only a quality. (this behavior is isolated to the `QuotedQualityCSV` implementation, the `QuotedCSV` implementation is not impacted).

Here's a breakdown of some of the test cases and why I feel that `valueLength` is superior.

Example calls to `parsedParam`

buffer | valueLength | paramName | paramValue | good/bad
------ | ----------- | --------- | ---------- | --------
`"="` | 0 | 0 | -1 | :x:
`"=="` | 0 | 0 | 1 | :x:
`"==="` | 0 | 0 | 1 | :x:
`";"` | 0 | -1 | -1 | :x:
`";;"` | 0 | -1 | -1 | :x:
`";;;"` | 0 | -1 | -1 | :x:
`";;="` | 0 | -1 | -1 | :x:
`";="` | 0 | -1 | -1 | :x:
`";=;"` | 0 | -1 | -1 | :x:
`"=;"` | 0 | -1 | -1 | :x:
`"=;;"` | 0 | -1 | -1 | :x:
`"=;="` | 0 | -1 | -1 | :x:
`"==;"` | 0 | -1 | -1 | :x:
`";=="` | 0 | -1 | 2 | :x:
`"=1"` | 0 | 0 | 1 | :x:
`"audio/*;q=0.2"` | 7 | 8 | 10 | :+1:
`"br;q=0.9"` | 2 | 3 | 5 | :+1:
`"br;q=0.99"` | 2 | 3 | 5 | :+1:
`"compress;q=0.5"` | 8 | 9 | 11 | :+1:
`"deflate;q=0.9"` | 7 | 8 | 10 | :+1:
`"gzip;q=0.9"` | 4 | 5 | 7 | :+1:
`"gzip;q=1.0"` | 4 | 5 | 7 | :+1:
`"identity;q=0.1"` | 8 | 9 | 11 | :+1:
`"identity;q=0.5"` | 8 | 9 | 11 | :+1:
`"one;q=0.5"` | 3 | 4 | 6 | :+1:
`"p=0.5"` | 0 | 0 | 2 | :x:
`"q=0"` | 0 | 0 | 2 | :x:
`"q="` | 0 | 0 | -1 | :x:
`"*;q=0.01"` | 1 | 2 | 4 | :+1:
`"q=;"` | 0 | -1 | -1 | :x:
`"*;q=0"` | 1 | 2 | 4 | :+1:
`"q=0.5"` | 0 | 0 | 2 | :x:
`"*/*;q=0.5"` | 3 | 4 | 6 | :+1:
`"q=x"` | 0 | 0 | 2 | :x:
`"sdch;q=0.7"` | 4 | 5 | 7 | :+1:
`"text/html;level=1"` | 9 | 10 | 16 | :+1:
`"text/html;level=2"` | 9 | 10 | 16 | :+1:
`"text/html;level=2;q=0.4"` | 9 | 18 | 20 | :+1:
`"text/html;q=0.7"` | 9 | 10 | 12 | :+1:
`"text/plain;format=flowed"` | 10 | 11 | 18 | :+1:
`"text/plain;q=0.5"` | 10 | 11 | 13 | :+1:
`"text/*;q=0.3"` | 6 | 7 | 9 | :+1:
`"text/x-dvi;q=0.8"` | 10 | 11 | 13 | :+1:
`"three;q=0.5"` | 5 | 6 | 8 | :+1:
`"three;x=y"` | 5 | 6 | 8 | :+1:
`"two;"` | 3 | -1 | -1 | :+1:
`"two;q=0.5"` | 3 | 4 | 6 | :+1:
`"value0.5;p=v"` | 8 | 9 | 11 | :+1:
`"value 0.5;p=v"` | 9 | 10 | 12 | :+1:
`"value0.5;p=v;q=0.5"` | 8 | 13 | 15 | :+1:
`"value 0.5;p=v;q="0.5""` | 9 | 14 | 16 | :+1:
`"value 0.5;p=v;q=0.5"` | 9 | 14 | 16 | :+1:
`"value 0.5;p="v  ;  q = \"0.5\"  ,  value 1.0 ""` | 9 | 10 | 12 | :+1:
`"valueBad;q=X"` | 8 | 9 | 11 | :+1:
`"value;p="v"` | 5 | 6 | 8 | :+1:


Example calls to `parsedValueAndParams`

buffer | valueLength | good/bad
------ | ----------- | --------
`";"` | 0 | :x:
`";;"` | 0 | :x:
`";;;"` | 0 | :x:
`";;="` | 0 | :x:
`";="` | 0 | :x:
`";=;"` | 0 | :x:
`";=="` | 0 | :x:
`"="` | 0 | :x:
`"=;"` | 0 | :x:
`"=;;"` | 0 | :x:
`"=;="` | 0 | :x:
`"=="` | 0 | :x:
`"==;"` | 0 | :x:
`"==="` | 0 | :x:
`"*"` | 1 | :+1:
`"=1"` | 0 | :x:
`"*/*"` | 3 | :+1:
`"aaaa"` | 4 | :+1:
`"audio/*"` | 7 | :+1:
`"audio/basic"` | 11 | :+1:
`"bbbb"` | 4 | :+1:
`"br"` | 2 | :+1:
`"cccc"` | 4 | :+1:
`"compress"` | 8 | :+1:
`"deflate"` | 7 | :+1:
`"gzip"` | 4 | :+1:
`"identity"` | 8 | :+1:
`"one"` | 3 | :+1:
`"p=0.5"` | 0 | :x:
`"q="` | 0 | :x:
`"q=;"` | 0 | :x:
`"q=0"` | 0 | :x:
`"q=0.5"` | 0 | :x:
`"q=x"` | 0 | :x:
`"sdch"` | 4 | :+1:
`"text/*"` | 6 | :+1:
`"text/html"` | 9 | :+1:
`"text/html;level=1"` | 9 | :+1:
`"text/html;level=2"` | 9 | :+1:
`"text/plain"` | 10 | :+1:
`"text/plain;format=flowed"` | 10 | :+1:
`"text/x-c"` | 8 | :+1:
`"text/x-dvi"` | 10 | :+1:
`"three"` | 5 | :+1:
`"three;x=y"` | 5 | :+1:
`"two"` | 3 | :+1:
`"value0.5;p=v"` | 8 | :+1:
`"value 0.5;p=v"` | 9 | :+1:
`"value 0.5;p="v  ;  q = \"0.5\"  ,  value 1.0 ""` | 9 | :+1:
`"value1.0"` | 8 | :+1:
`"value 1.0"` | 9 | :+1:
`"valueBad"` | 8 | :+1:
`"value;p="v"` | 5 | :+1:

gregw commented:
> @gregw I reworked some of the points you highlighted with regards to buffer.toString() and valueLength.

Are you pushed? I can't see any changes?

I really don't understand how valueLength is going to work.  You have to parse the string to identify what the value length is before passing it into the parser?

gregw reviewed (changes requested):
I do not believe we should be changing our qsv parsing in 12.0.x at all.
We should just fix the NPE and return a null locale.

In 12.1.x, I think we need to cleanup QSV and CSV parsing to bring them in line with 9110.  

There is an argument to be made that a badly formed QSV list should result in a 400 Bad request response.   Perhaps we need a compatibility mode to support any changes to a stricter QSV parser.

Either way, that class and associated classes are showing some poor evolution and need a bit of a cleanup.

@joakime Can you change this PR to a minimal change to avoid the NPE and them move your new tests for bad QSVs to a new branch for 12.1.x and we can think about how to fix it better there.

joakime commented:
> Are you pushed? I can't see any changes?

The changes were pushed.  See: https://github.com/jetty/jetty.project/pull/13604/commits/6fc293071cb406d4c006c2349d9de8edde35344f

joakime commented:
> @joakime Can you change this PR to a minimal change to avoid the NPE and them move your new tests for bad QSVs to a new branch for 12.1.x and we can think about how to fix it better there.

This is what I originally did, prevent the NPE in getLocale() in every ee# implementation, and core instead.
It's actually a bigger PR that way, way more lines of code changed.

If that's what you want, i'll prepare it that way.  But be prepared for the changes.

joakime committed:
Expand RequestTest.testAcceptableLocales tests

gregw reviewed (approved):
This is a good way of doing it.
But in 12.1.x it will be better to filter out illegal values from the QCSV, but I still think we need to consider a compatibility mode to switch between ignore and reject.
